kong/db/errors.lua

```diff
@@ -510,4 +516,458 @@ function _M:invalid_unique_global(name)
+  --       message = "failed conditional validation given value of field 'protocol'",
+  --     },
+  --     {
+  --       type = "field"
+  --       field = "path",
+  --       message = "value must be null",
+  --     }
+  --   }
+  -- }
+  -- ```
+  --
+  -- Nested foreign relationships are hoisted up to the top level, so
+  -- given the following input:
+  --
+  -- ```lua
+  -- {
+  --   services = {
+  --     name = "matthew",
+  --     url = "http:/127.0.0.1:80/",
+  --     routes = {
+  --       {
+  --         name = "joshua",
+  --         protocols = { "nope" },            -- invalid protocol
+  --       }
+  --     },
+  --     plugins = {
+  --       {
+  --         name = "i-am-not-a-real-plugin",   -- nonexistent plugin
+  --         config = {
+  --           foo = "bar",
+  --         },
+  --       },
+  --       {
+  --         name = "http-log",
+  --         config = {},                       -- missing required field(s)
+  --       },
+  --     },
+  --   }
+  -- }
+  -- ```
+  -- ... the output error array will have three entries, one for the route,
+  -- and one for each of the plugins.
+  --
+  -- Errors for record fields and nested schema properties are rolled up and
+  -- added to their parent entity, with the full path to the property
+  -- represented as a period-delimited string:
+  --
+  -- ```lua
+  -- {
+  --   entity_type = "plugin",
+  --   entity_name = "http-log",
+  --   entity = {
+  --     name = "http-log",
+  --     config = {
+  --       -- empty
+  --     },
+  --   },
+  --   errors = {
+  --     {
+  --       field = "config.http_endpoint",
+  --       message = "missing host in url",
+  --       type = "field"
+  --     }
+  --   },
+  -- }
+  -- ```
+  function flatten_errors(input, err_t)
+    local flattened = {}
+
+    for entity_type, section_errors in pairs(err_t) do
+      if type(section_errors) ~= "table" then
+        log(WARN, "failed to resolve errors for ", entity_type)
+        goto next_section
+      end
+
+      local entities = input[entity_type]
+
+      if type(entities) ~= "table" then
+        log(WARN, "failed to resolve errors for ", entity_type)
+        goto next_section
+      end
+
+      for idx, errs in pairs(section_errors) do
+        local entity = entities[idx]
+
+        if type(entity) == "table" then
+          add_entity_errors(entity_type, entity, errs, flattened)
+
+        else
+          log(WARN, "failed to resolve errors for ", entity_type, " at ",
+                    "index '", idx, "'")
+        end
+      end
+
+      ::next_section::
+    end
+
+    return flattened
+  end
+end
+
+
+function _M:declarative_config_flattened(err_t, input)
+  if type(err_t) ~= "table" then
+    error("err_t must be a table", 2)
+  end
+
+  if type(input) ~= "table" then
+    error("err_t input is nil or not a table", 2)
+  end
+
+  local flattened = flatten_errors(input, err_t)
+
+  err_t = self:declarative_config(err_t)
+
+  err_t.flattened = flattened
```

Comment:
*Feedback wanted!*

Is `flattened` the attribute name that we want to stick with? I feel kinda "meh" about it myself. Keep in mind we must maintain backwards compatibility with the old response shape.

Given this response:

```json
{
  "message": "declarative config is invalid: {services={{host=\"required field missing\",port=\"value should be between 0 and 65535\"}}}",
  "code": 14,
  "fields": {
    "services": [
      {
        "host": "required field missing",
        "port": "value should be between 0 and 65535"
      }
    ]
  },
  "name": "invalid declarative configuration"
}
```

...the new "shape" will look like this:

```json
{
  "message": "declarative config is invalid: {services={{host=\"required field missing\",port=\"value should be between 0 and 65535\"}}}",
  "code": 14,
  "fields": {
    "services": [
      {
        "host": "required field missing",
        "port": "value should be between 0 and 65535"
      }
    ]
  },
  "name": "invalid declarative configuration",
  "flattened": [
    {
      "entity_type": "service",
      "entity_name": "jeff",
      "entity_tags": [
        "jeff-is-cool"
      ],
      "entity": {
        "name": "jeff",
        "protocol": "http",
        "port": -1,
        "tags": [
          "jeff-is-cool"
        ]
      },
      "errors": [
        {
          "type": "field",
          "field": "host",
          "message": "required field missing"
        },
        {
          "type": "field",
          "field": "port",
          "message": "value should be between 0 and 65535"
        }
      ]
    }
  ]
}
```

Similar options could be...
  - `errors`
  - `entities`
  - `entity_errors`
  - `flattened_errors`

Reply:
It was an arbitrary name I came up with without a whole lot of though, to contrast within the traditional "nested" structure of `fields`.

I'm not particularly attached to it and it's trivial to adjust to a change on the controller side, so if you have a name you prefer. go for it.

Reply:
Naming is hard, but we all know Jeff is cool
![](https://media.tenor.com/KVosJDEqP_UAAAAC/bo-burnham-inside.gif)

I prefer `flattened_errors` myself.

Reply:
I think I'll go with `flattened_errors`, if only because `flattened` is a word that is used all over the place in relation to dbless configuration/schema, so this helps sets it apart ever so slightly.